File: frontend/src/services/bookService.ts
import api from './api';

export interface Book {
  id?: number;
  title: string;
  author: string;
  isbn: string;
  year: number;
  copies: number;
}

export const bookService = {
  // Get all books
  getAllBooks: async (): Promise<Book[]> => {
    const response = await api.get('/api/books');
    return response.data;
  },

  // Get a single book by ID
  getBookById: async (id: number): Promise<Book> => {
    const response = await api.get(`/api/books/${id}`);
    return response.data;
  },

  // Create a new book
  createBook: async (book: Omit<Book, 'id'>): Promise<Book> => {
    const response = await api.post('/api/books', book);
    return response.data;
  },

  // Update an existing book
  updateBook: async (id: number, book: Book): Promise<Book> => {
    const response = await api.put(`/api/books/${id}`, book);
    return response.data;
  },

  // Delete a book
  deleteBook: async (id: number): Promise<void> => {
    await api.delete(`/api/books/${id}`);
  },

  // Search books
  searchBooks: async (query: string): Promise<Book[]> => {
    const response = await api.get(`/api/books/search?query=${encodeURIComponent(query)}`);
    return response.data;
  }
}; 

File: api/src/main/java/com/library/api/service/BookService.java
package com.library.api.service;

import com.library.api.model.Book;
import java.util.List;
import java.util.Optional;

public interface BookService {
    // CRUD Operations
    Book saveBook(Book book);

    List<Book> getAllBooks();

    Optional<Book> getBookById(Long id);

    Optional<Book> getBookByIsbn(String isbn);

    Book updateBook(Long id, Book book);

    void deleteBook(Long id);

    // Business Logic Operations
    List<Book> findBooksByTitle(String title);

    List<Book> findBooksByAuthor(String author);

    List<Book> findBooksByYear(int year);

    List<Book> findAvailableBooks();

    boolean isBookAvailable(String isbn);

    void updateBookCopies(String isbn, int copies);

    // Search Operations
    List<Book> searchBooks(String query);
}